RWAHS-502 Search within containers

leftclickben opened the pull request:
Fixes both issues relating to searching elements within containers:
- The query builder only lists containers themselves, not their children, which is where the actual data is.
- CA only indexes the ID for list elements nested under containers, not the text values.  (For "top-level" list elements, i.e. not in a container, the text value is indexed).

Commit:
Make the query builder display container children and not the containers themselves.

+ The query builder was displaying a filter for container elements, which do
  not directly store data (and therefore useless to search on).
+ The query builder was *not* displaying a filter for the children of container
  elements, which is where the data is stored (and therefore should be used
  in searches).
+ This changes the query builder to display the child elements and not the
  containers themselves.

Commit:
Index text values of list elements nested inside container elements.

* Yes, it is a good idea (re: comment in code that has been removed
  in this commit).
* For example, if there is a container "Creator", which has a "Creator
  Type" sub-element and a "Creator Name" sub-element, this change
  allows the actual text value of the "Creator Type" to be indexed.

Commit:
DRY.

Commit:
Fix search indexing for nested, non-list values.

+ Previous changes broke the indexing of nested (in a
  container) elements that are not of list type (or various
  other types that are handled in the same way as lists).
+ This fixes it by not clobbering the passed-in value for
  `$va_attributes`, and only operating on the value for
  the specific child being indexed, not all values within the
  container (added `if` condition).

leftclickben commented:
https://www.youtube.com/watch?v=04854XqcfCY

leftclickben commented:
Dammit... Will fix tests tomorrow.

kehh commented:
Well you are anyway (your previous comment)

leftclickben commented:
Nah we were literally 2 lines away from making the lists in containers work. Definitely a combined effort.

kehh commented on app/lib/core/Search/SearchIndexer.php:
```diff
@@ -1437,7 +1408,7 @@ private function _indexAttribute($pt_subject, $pn_row_id, $pm_element_code_or_id
 
 						$va_children_ids = $pt_subject->getHierarchyAsList($pn_row_id, array('idsOnly' => true, 'includeSelf' => false));
 
-						if (!$pb_reindex_mode && is_array($va_children_ids) && sizeof($va_children_ids) > 0) {
```
Seems suspicious - why remove this check?

leftclickben replied:
`$pb_reindex_mode` was undefined, which is falsy, so `!$pb_reindex_mode` is always going to be true, therefore this check can be removed.  Good pick up though :-)

Commit:
Add safety check to ensure array exists before iterating over it.

+ Fixes unit tests.
+ Strictly this shouldn't be needed, because containers will always have an
  array of child elements (even if the array is empty).
+ However it is safe to keep here, and fixing the actual tests is more effort.

Commit:
Add missing typecasts.

leftclickben commented:
Assuming the tests pass in travis (they pass locally now) this should be good to go.
